website-cleaner/main.go

```diff
@@ -0,0 +1,297 @@
+	".lang-selector", ".language", "#language-selector", ".social", ".social-media",
+	".social-links", "#social", ".menu", ".navigation", "#nav", ".breadcrumbs",
+	"#breadcrumbs", "#search-form", ".search", "#search", ".share", "#share",
+	".widget", "#widget", ".cookie", "#cookie",
+}
+
+func main() {
+
+	input := os.Getenv("INPUT")
+	output := os.Getenv("OUTPUT")
+
+	logOut := logrus.New()
+	logOut.SetOutput(os.Stdout)
+	logOut.SetFormatter(&logrus.JSONFormatter{})
+	logErr := logrus.New()
+	logErr.SetOutput(os.Stderr)
+
+	ctx := context.Background()
+	gptscriptClient, err := gptscript.NewGPTScript()
+	if err != nil {
+		logOut.WithError(fmt.Errorf("failed to create gptscript client, error: %v", err)).Error()
+		os.Exit(0)
+	}
+
+	if input == "" {
+		logOut.WithError(fmt.Errorf("input is empty")).Error()
+		os.Exit(0)
+	}
+
+	if output == "" {
+		logOut.WithError(fmt.Errorf("output is empty")).Error()
+		os.Exit(0)
+	}
+
+	inputFile, err := gptscriptClient.ReadFileInWorkspace(ctx, input)
+	if err != nil {
+		logOut.WithError(fmt.Errorf("failed to read input file %q: %v", input, err)).Error()
+		os.Exit(0)
+	}
+
+	originalSize := len(inputFile)
+
+	doc, err := goquery.NewDocumentFromReader(bytes.NewBuffer(inputFile))
+	if err != nil {
+		logOut.WithError(fmt.Errorf("failed to parse html input: %v", err)).Error()
+		os.Exit(0)
+	}
+
+	// Clean HTML programmatically
+	for _, tag := range tagsToRemove {
+		doc.Find(tag).Remove()
+	}
+
+	// transform to Markdown
+	converter := md.NewConverter(md.WithPlugins(base.NewBasePlugin(), commonmark.NewCommonmarkPlugin()))
+	html, err := doc.Html()
+	if err != nil {
+		logOut.WithError(fmt.Errorf("failed to get html from document: %v", err)).Error()
+		os.Exit(0)
+	}
+
+	sanitizedHTMLSize := len(html)
+
+	markdown, err := converter.ConvertString(html)
+	if err != nil {
+		logOut.WithError(fmt.Errorf("failed to convert html to markdown: %v", err)).Error()
+		os.Exit(0)
+	}
+
+	markdownSize := len(markdown)
+
+	if markdownSize < 100_000 {
+		logErr.Info("Markdown content is less than 100,000 characters -> sending to LLM for further cleanup")
+
+		prompt := "The following content is a scraped webpage converted to markdown. Please remove any content that came from the website header, footer, or navigation. The output should focus on just the main content body of the page. Maintain the markdown format, including any links or images.\n\n" + markdown
+
+		url := fmt.Sprintf("%s/chat/completions", os.Getenv("OPENAI_BASE_URL"))
+
+		headers := map[string]string{
+			"Content-Type":  "application/json",
+			"Authorization": "Bearer " + os.Getenv("OPENAI_API_KEY"),
+		}
+
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, 10*time.Minute)
+		defer cancel()
+
+		model := os.Getenv("OTTO8_DEFAULT_LLM_MINI_MODEL")
+		if model == "" {
+			logOut.WithError(fmt.Errorf("OTTO8_DEFAULT_LLM_MINI_MODEL is empty")).Error()
+			os.Exit(0)
+		}
```

Comment:
Darren explicitly asked that we default to something reasonable in this case. I think `gpt-4o-mini` would be good here.